### PR TITLE
Add LlamaIndex as a Datastore

### PR DESCRIPTION
**Pull request by Disiok**
### Summary
* Initial integration for using [LlamaIndex](https://github.com/jerryjliu/llama_index) as a Datastore

### Details
* Add `LlamaDataStore` implementing `DataStore`
* Update documentation
* Add `llama-index = "0.5.4"` dependency

### CODE REVIEW

File: tests/datastore/providers/llama/test_llama_datastore.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+    assert len(query_0_results) == 1
+    assert len(query_1_results) == 2
+    
+    # NOTE: this is the correct behavior
+    # assert query_0_results[0].id == 'first-doc-4'
+    # assert query_1_results[0].id == 'first-doc-5'
+    # assert query_1_results[1].id == 'first-doc-4'
+
+    # TODO: this is current buggy behavior, to be fixed in 0.5.2
```
Comment (Disiok) [inline tests/datastore/providers/llama/test_llama_datastore.py:81]: Pending: https://github.com/jerryjliu/llama_index/pull/977

File: README.md [hunk 2]
```diff
@@ -257,6 +257,14 @@ For more detailed instructions on setting up and using each vector database prov
 
 [Redis](https://redis.com/solutions/use-cases/vector-database/) is a real-time data platform suitable for a variety of use cases, including everyday applications and AI/ML workloads. It can be used as a low-latency vector engine by creating a Redis database with the [Redis Stack docker container](/examples/docker/redis/docker-compose.yml). For a hosted/managed solution, [Redis Cloud](https://app.redislabs.com/#/) is available. For detailed setup instructions, refer to [`/docs/providers/redis/setup.md`](/docs/providers/redis/setup.md).
 
+
+#### LlamaIndex
+
+[LlamaIndex](https://github.com/jerryjliu/llama_index) is a central interface to connect your LLM's with external data.
+It provides a suite of in-memory indices over your unstructured and structured data for use with ChatGPT.
+It is light-weight and easy-to-use. All you need to do is specifying a few environment variables (optionally point to an existing saved Index json file), no additional deployment necessary.
```
Comment (isafulf) [inline README.md:265]: Maybe add a sentence contrasting this to the other vector db providers/ describing the benefits of using LlamaIndex?

Also for transparency, add a sentence mentioning that metadata filters in queries are not yet supported.

File: docs/providers/llama/setup.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+| `LLAMA_QUERY_KWARGS_JSON_PATH`         | Optional | Path to saved query kwargs json file                                      | None               |
+| `LLAMA_RESPONSE_MODE`           | Optional | Response mode for query                                            | `no_text`          | 
+
+
+**Different Index Types**
+By default, we use a `GPTSimpleVectorIndex` to store document chunks in memory, 
+and retrieve top-k nodes by embedding similarity.
+Different index types are optimized for different data and query use-cases.
+See this guide on [How Each Index Works](https://gpt-index.readthedocs.io/en/latest/guides/primer/index_guide.html) to learn more.
+You can configure the index type via the `LLAMA_INDEX_TYPE`, see [here](https://gpt-index.readthedocs.io/en/latest/reference/indices/composability_query.html#gpt_index.data_structs.struct_type.IndexStructType) for the full list of accepted index type identifiers.
```
Comment (isafulf) [inline docs/providers/llama/setup.md:31]: Could you add a sentence on how to run the tests?

**Review by isafulf — COMMENTED** [review]
Cool!! Left a couple comments.

Could you also update this file please:
https://github.com/openai/chatgpt-retrieval-plugin/blob/main/docs/deployment/removing-unused-dependencies.md

Thanks :)

File: README.md (hunk 2 above)
Comment (Disiok) [reply README.md:265]: Great suggestions. Will revise the docs

File: docs/providers/llama/setup.md (hunk 3 above)
Comment (Disiok) [reply docs/providers/llama/setup.md:31]: Sure, will revise asap.

**Review by isafulf — APPROVED** [review]
Great! Merging

**Pull request merged**